src/timely-util/src/reclock.rs

```diff
@@ -0,0 +1,609 @@
+            if PartialOrder::less_than(&replayer.upper(), &upper) {
+                accepted_bindings.sort_unstable_by(|a, b| a.1.cmp(&b.1));
+                // The times are totally ordered so we can binary search to find the prefix that is
+                // not beyond the upper and extract it into a batch.
+                let idx = accepted_bindings.partition_point(|(_, t, _)| !upper.less_equal(t));
+                replayer.append_batch(accepted_bindings.drain(0..idx), upper.to_owned());
+            }
+
+            // Accept new data
+            while let Some(event) = puller.pull() {
+                match event.as_mut() {
+                    Event::Progress(changes) => {
+                        source_upper.update_iter(changes.drain(..));
+                    }
+                    Event::Messages(_, data) => {
+                        data.sort_unstable_by(|a: &(D1, T, R), b| a.1.cmp(&b.1));
+                        let batch = Batch::from_iter(data.drain(..));
+                        source_batches.push_back(batch);
+                    }
+                }
+            }
+
+            // Reclock accepted data
+            let mut output = output.activate();
+            while let Some((frontier, cap)) = replayer.active_binding() {
+                let into_t = cap.time();
+                source_batches.retain_mut(|batch| {
+                    batch.chains.retain_mut(|chain| {
+                        // A chain is a set of mutually comparable elements so we can binary search
+                        // to find the prefix that is not beyond the frontier and extract it.
+                        let idx = chain.partition_point(|(_, t, _)| !frontier.less_equal(t));
+                        let data = chain
+                            .drain(0..idx)
+                            .map(|(d, t, r)| (logic(d, t), into_t.clone(), r));
+                        container.extend(data);
+                        // Retain non-empty chains
+                        !chain.is_empty()
+                    });
+                    // Retain non-empty batches
+                    !batch.chains.is_empty()
+                });
+                if !container.is_empty() {
+                    output.session(cap).give_container(&mut container);
+                }
+                // If we won't receive any more data for this binding we can go to the next one
+                if PartialOrder::less_equal(&frontier, &source_upper.frontier()) {
+                    replayer.step();
+                } else {
+                    // Otherwise yield and wait for more data
+                    break;
+                }
+            }
+        }
+    });
+
+    (Box::new(pusher), reclocked.as_collection())
+}
+
+/// A struct that reveals the frontiers encoded in a remap binding collection in time order.
+struct BindingReplay<FromTime, IntoTime: Timestamp> {
+    /// The upper frontier of bindings received so far
```

Comment:
Nit: This is only a frontier for totally-ordered times, the comment could be clearer.

Reply:
Yeah, I will move the `TotalOrder` bound on the struct too and spell this out in the doc string